pkg/cache/metrics_test.go

```diff
@@ -0,0 +1,96 @@
+//go:generate mockgen --build_flags=--mod=mod -package cache -destination ./mocks_test.go github.com/rancher/lasso/pkg/client SharedClientFactory
+package cache
+
+import (
+	"context"
+	"strings"
+	"testing"
+	"time"
+
+	gomock "github.com/golang/mock/gomock"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+	"github.com/rancher/lasso/pkg/client"
+	"github.com/rancher/lasso/pkg/metrics"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+)
+
+func setupMockSharedClientFactory(t *testing.T, gvr schema.GroupVersionResource, gvk schema.GroupVersionKind) client.SharedClientFactory {
+	t.Helper()
+	cf := NewMockSharedClientFactory(gomock.NewController(t))
+
+	testClient := &client.Client{}
+	scheme := runtime.NewScheme()
+	if err := clientgoscheme.AddToScheme(scheme); err != nil {
+		t.Fatal(err)
+	}
+
+	cf.EXPECT().ResourceForGVK(gvk).DoAndReturn(func(gvk schema.GroupVersionKind) (schema.GroupVersionResource, bool, error) {
+		return gvr, true, nil
+	}).AnyTimes()
+	cf.EXPECT().ForKind(gvk).DoAndReturn(func(gvk schema.GroupVersionKind) (*client.Client, error) {
+		return testClient, nil
+	}).AnyTimes()
+	cf.EXPECT().NewObjects(gvk).DoAndReturn(func(gvk schema.GroupVersionKind) (runtime.Object, runtime.Object, error) {
+		obj, err := scheme.New(gvk)
+		if err != nil {
+			return nil, nil, err
+		}
+		objList, err := scheme.New(schema.GroupVersionKind{
+			Group:   gvk.Group,
+			Version: gvk.Version,
+			Kind:    gvk.Kind + "List",
+		})
+
+		return obj, objList, err
+	}).AnyTimes()
+	cf.EXPECT().ForResourceKind(gvr, gvk.Kind, true).DoAndReturn(func(gvr schema.GroupVersionResource, kind string, namespaced bool) *client.Client {
+		return testClient
+	})
+
+	return cf
+}
+
+func Test_sharedCacheFactory_metrics_collection(t *testing.T) {
+	gvr := corev1.SchemeGroupVersion.WithResource("configmaps")
+	gvk := corev1.SchemeGroupVersion.WithKind("ConfigMap")
+
+	cf := setupMockSharedClientFactory(t, gvr, gvk)
+	scf := NewSharedCachedFactory(cf, &SharedCacheFactoryOptions{MetricsCollectionPeriod: time.Millisecond * 200}).(*sharedCacheFactory)
+	_, err := scf.ForKind(gvk)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
```

Comment:
You're calling `cancel`  later in the code?

Reply:
Yes, cancelling the same context twice is harmless, and eliminates the chance of the first assertion failing, so that `t.Fatal` will make the `cancel()` unreachable.
Also, it kind of hurts my eyes not seeing `defer cancel()` after creating the context 😅 I used `t.Cleanup` though to express it's actual cleanup.